Dockerfile

```diff
@@ -46,18 +46,20 @@ RUN --mount=type=cache,target=/root/.cache/go-build \
 ARG package=.
 ARG ARCH
 ARG ldflags
+ARG outbin="manager"
+ENV execbin=$outbin
 
 # Do not force rebuild of up-to-date packages (do not use -a) and use the compiler cache folder
 RUN --mount=type=cache,target=/root/.cache/go-build \
     --mount=type=cache,target=/go/pkg/mod \
     CGO_ENABLED=0 GOOS=linux GOARCH=${ARCH} \
     go build -trimpath -ldflags "${ldflags} -extldflags '-static'" \
-    -o manager ${package}
+    -o ${outbin} ${package}
 
 # Production image
 FROM gcr.io/distroless/static:nonroot
 WORKDIR /
-COPY --from=builder /workspace/manager .
+COPY --from=builder /workspace/${outbin} .
 # Use uid of nonroot user (65532) because kubernetes expects numeric user when applying pod security policies
 USER 65532
-ENTRYPOINT ["/manager"]
+ENTRYPOINT exec ${execbin}
```

Comment:
I am not a 100% sure how the clusterctl containers will be used but when we build the container like this we will not be able to run the clusterctl container as it fails with this: 
```
$ docker run --rm  gcr.io/kakaraparthy-devel/clusterctl-amd64:dev version

docker: Error response from daemon: failed to create shim: OCI runtime create failed: container_linux.go:380: starting container process caused: exec: "/bin/sh": stat /bin/sh: no such file or directory: unknown.
```

I think the problem is combination of using the build args to pass the binary name and the way the `ENTRYPOINT` is constructed. 

It might be cleaner to have a separate dockerfile for building clusterctl and a separate dockerfile that we can use for all the managers. 

Thoughts on this @fabriziopandini ?

Reply:
>I think using one Dockerfile should be fine. Why are we using exec instead of ["/${execbin}"] ?

I think this doesn't work too. I vaguely remember trying something like this and experiencing errors. IIRC it is because `[/manager]` is the Exec form and in this form a shell is not used to execute and therefore no environment variable substitutions are done because there is no shell to do it. 

May be @sachinkumarsingh092 can experiment with this and report any findings.

Reply:
Yes. Another way to do this in exec for is ENTRYPOINT [ "sh", "-c", "$execbin" ], but this relies on the assumption that a `sh` shell is already present.
From the [docker docs](https://docs.docker.com/engine/reference/builder/#entrypoint):
```
Unlike the shell form, the exec form does not invoke a command shell. This means that normal shell processing does not happen. For example, ENTRYPOINT [ "echo", "$HOME" ] will not do variable substitution on $HOME. 
If you want shell processing then either use the shell form or execute a shell directly, for example: ENTRYPOINT [ "sh", "-c", "echo $HOME" ]. When using the exec form and executing a shell directly, as in the case for the shell form, it is the shell that is doing the environment variable expansion, not docker.
```

Reply:
So this mean we would have to use a base image with a shell just to make the name of the binary parameterizable?

If that's the case I think we should use a separate dockerfile.

Using the same binary name for controllers (aka manager) and clusterctl is probably not a good idea.

Reply:
I think the distroless image that we use does not come with a shell. Lets do with separate Dockerfiles.